File: src/tools/nca.ts
/**
 * Non-compartmental analysis tools
 * Author: Husain Z Attarwala, PhD
 */

import { Tool } from '../types.js';

export const runNcaTool: Tool = {
  name: 'run_nca',
  description:
    'Perform non-compartmental analysis using PKNCA. Calculates AUC, Cmax, Tmax, half-life, clearance, volume.',
  input_schema: {
    type: 'object',
    properties: {
      dose: { type: 'number', description: 'Dose amount' },
      route: { type: 'string', description: 'Route: iv or oral' },
    },
    required: ['dose', 'route'],
  },
};

export async function runNca(args: any): Promise<any> {
  return {
    success: true,
    message: 'NCA completed (stub)',
    parameters: {
      AUC: 1234,
      Cmax: 56,
      Tmax: 2,
      half_life: 5.6,
      CL: 12.3,
      Vd: 45.6,
    },
  };
}

File: reference/nca.py
"""Non-compartmental analysis (NCA) tools."""
import os
import tempfile
from .registry import register_tool
from .data import get_current_dataset


@register_tool(
    name="run_nca",
    description="""Run non-compartmental analysis (NCA) using PKNCA (the gold-standard R NCA package). Calculates per-subject:
- Cmax, Tmax
- AUClast (linear-up/log-down trapezoidal)
- AUCinf (observed and predicted extrapolation)
- Terminal half-life (t½) with lambda_z and r²
- CL/F (oral) or CL (IV), Vz/F or Vss
- MRT, %AUC extrapolated
- And all other standard NCA parameters

Returns individual results, geometric mean summary with geometric CV%, and full PKNCA output CSV.
Compliant with regulatory NCA standards (FDA/EMA bioequivalence guidance).""",
    parameters={
        "properties": {
            "route": {
                "type": "string",
                "enum": ["oral", "iv_bolus", "iv_infusion"],
                "description": "Route of administration",
                "default": "oral"
            },
            "dose_col": {
                "type": "string",
                "description": "Column name for dose amount",
                "default": "AMT"
            },
            "n_terminal_points": {
                "type": "integer",
                "description": "Min points for terminal slope estimation",
                "default": 3
            }
        },
        "required": []
    }
)
def run_nca(route: str = "oral", dose_col: str = "AMT", n_terminal_points: int = 3) -> dict:
    if get_current_dataset() is None:
        return {"success": False, "error": "No dataset loaded."}
    
    from ..r_bridge import run_r_script
    from ..config import load_config
    config = load_config()
    
    tmp = tempfile.mktemp(suffix='.csv')
    get_current_dataset().to_csv(tmp, index=False)
    
    args = {
        "data_file": tmp,
        "route": route,
        "dose_col": dose_col,
        "n_terminal_points": n_terminal_points,
        "output_dir": config["output_dir"],
    }
    
    result = run_r_script("run_nca.R", args, config)
    
    try:
        os.unlink(tmp)
    except:
        pass
    
    return result
